WordPress/Classes/Login/WordPressDotComAuthenticator.swift

```diff
@@ -0,0 +1,107 @@
+            .appending(queryItems: [
+                URLQueryItem(name: "client_id", value: clientId),
+                URLQueryItem(name: "redirect_uri", value: redirectURI),
+                URLQueryItem(name: "response_type", value: "code"),
+                URLQueryItem(name: "scope", value: "global"),
+            ])
+
+        let callbackURL: URL = try await withCheckedThrowingContinuation { continuation in
+            DispatchQueue.main.async {
+                let provider = ContextProvider(window: viewController.view.window ?? UIWindow())
+                let session = ASWebAuthenticationSession(url: authorizeURL, callbackURLScheme: "x-wordpress-app") { url, error in
+                    if let url {
+                        continuation.resume(returning: url)
+                    } else {
+                        DDLogWarn("Error from authentication session: \(String(describing: error))")
+                        continuation.resume(throwing: Error.cancelled)
+                    }
+                }
+                session.presentationContextProvider = provider
+                session.start()
+            }
+        }
+
+        return try await handleAuthorizeCallbackURL(callbackURL, clientId: clientId, clientSecret: clientSecret, redirectURI: redirectURI)
+    }
+
+    private func handleAuthorizeCallbackURL(
+        _ url: URL,
+        clientId: String,
+        clientSecret: String,
+        redirectURI: String
+    ) async throws -> String {
+        guard let query = URLComponents(url: url, resolvingAgainstBaseURL: true)?.queryItems,
+              let code = query.first(where: { $0.name == "code" })?.value else {
+            throw Error.invalidCallbackURL
+        }
+
+        var tokenRequest = URLRequest(url: URL(string: "https://public-api.wordpress.com/oauth2/token")!)
+        tokenRequest.httpMethod = "POST"
+        let parameters = [
+            "grant_type": "authorization_code",
+            "client_id": clientId,
+            "client_secret": clientSecret,
+            "redirect_uri": redirectURI,
+            "code": code,
+        ]
+
+        do {
+            tokenRequest = try URLEncodedFormParameterEncoder().encode(parameters, into: tokenRequest)
+        } catch {
+            DDLogError("Failed to encode token request")
+            throw Error.obtainAccessToken
+        }
+
+        do {
+            let urlSession = URLSession(configuration: .default)
+            let (data, _) = try await urlSession.data(for: tokenRequest)
+
+            struct Response: Decodable {
+                var accessToken: String
+            }
+
+            let decoder = JSONDecoder()
+            decoder.keyDecodingStrategy = .convertFromSnakeCase
+            let token = try decoder.decode(Response.self, from: data).accessToken
+
+            return token
+        } catch {
+            DDLogError("Failed to parse token request response: \(error)")
+            throw Error.obtainAccessToken
```

Comment:
I have added more error cases in https://github.com/wordpress-mobile/WordPress-iOS/pull/23521/commits/ffc7d5203091e632e9b8f056e16073ba4f171604.

BTW, I didn't throw the original errors because I'd like to use typed throw once it's available.